lib/CodeGen/CGExprScalar.cpp

```diff
@@ -2342,8 +2342,17 @@ llvm::Value *ScalarExprEmitter::EmitIncDecConsiderOverflowBehavior(
 llvm::Value *
 ScalarExprEmitter::EmitScalarPrePostIncDec(const UnaryOperator *E, LValue LV,
                                            bool isInc, bool isPre) {
-
   QualType type = E->getSubExpr()->getType();
+
+  // Add a dynamic check for pointer arithmetic of null pointers.
+  if (type->isCheckedPointerType() || type->isCheckedArrayType()) {
+    LValueBaseInfo BaseInfo;
+    TBAAAccessInfo TBAAInfo;
+    Address Addr = CGF.EmitPointerWithAlignment(E->getSubExpr(),
```

Comment:
This will result in E's subexpression being evaluated twice.  This would be a problem for something where the subexpression might have a side-effect, such as `(a[f()])++`.   I think you need to push the insertion of the check into the cases below.   The only interesting cases are those where a pointer type can occur.